src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -362,96 +364,146 @@ CheckDistributedTable(Oid relationId)
+		insertedShardPlacements = lappend(insertedShardPlacements, shardPlacement);
+
+		placementsInserted++;
+		if (placementsInserted >= replicationFactor)
 		{
-			const RelayFileState shardState = FILE_FINALIZED;
-			const uint64 shardSize = 0;
-
-			InsertShardPlacementRow(shardId, INVALID_PLACEMENT_ID, shardState, shardSize,
-									nodeName, nodePort);
-			placementsCreated++;
+			break;
 		}
-		else
+	}
+
+	return insertedShardPlacements;
+}
+
+
+/*
+ * CreateShardsOnWorkers creates shards on worker nodes given the shard placements
+ * as a parameter. Function opens connections in transactional way if the caller
+ * doesn't intentionally need non-transactional one. If the caller needs an
+ * exclusive connection (in case of distributing local table with data on it)
+ * or creating shards in a transaction, per placement connection is opened for
+ * each placement.
+ */
+void
+CreateShardsOnWorkers(Oid distributedRelationId, List *shardPlacements,
+					  bool useExclusiveConnection, bool createInSeparateTransaction,
+					  bool colocatedShard)
+{
+	char *placementOwner = TableOwner(distributedRelationId);
+	bool includeSequenceDefaults = false;
+	List *ddlCommandList = GetTableDDLEvents(distributedRelationId,
+											 includeSequenceDefaults);
+	List *foreignConstraintCommandList = GetTableForeignConstraintCommands(
+		distributedRelationId);
+	List *claimedConnectionList = NIL;
+	ListCell *connectionCell = NULL;
+	ListCell *shardPlacementCell = NULL;
+
+	foreach(shardPlacementCell, shardPlacements)
+	{
+		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(shardPlacementCell);
+		uint64 shardId = shardPlacement->shardId;
+		ShardInterval *shardInterval = LoadShardInterval(shardId);
+		MultiConnection *connection = NULL;
+		int shardIndex = -1;
+
+		if (colocatedShard)
 		{
-			ereport(WARNING, (errmsg("could not create shard on \"%s:%u\"",
-									 nodeName, nodePort)));
+			shardIndex = ShardIndex(shardInterval);
 		}
 
-		if (placementsCreated >= replicationFactor)
+		/*
+		 * Shards must be created within separate transaction if it is created
+		 * for the non-hash distributed tables, because users can copy data from
+		 * workers to these tables. If shards do not created within separate
+		 * transaction, worker node can not see the uncommitted shard creation
+		 * transaction which is the case show up in the copy.
+		 */
+		if (!createInSeparateTransaction)
 		{
-			break;
+			BeginOrContinueCoordinatedTransaction();
+			connection = GetPlacementConnection(FOR_DDL, shardPlacement,
+												placementOwner);
+			if (IsTransactionBlock() || !useExclusiveConnection)
+			{
+				ClaimConnectionExclusively(connection);
+				claimedConnectionList = lappend(claimedConnectionList, connection);
+			}
+
+			RemoteTransactionBeginIfNecessary(connection);
+			MarkRemoteTransactionCritical(connection);
 		}
+		else
+		{
+			connection = GetPlacementConnection(FORCE_NEW_CONNECTION, shardPlacement,
```

Comment:
Correct. There's probably not as much value in failing over if a SQL command errors out, since that only happens if there is something wrong with the database (rare, and probably not something we want to hide) or the connection fails immediately after establishing it (very rare). If not doing it keeps the code simpler then that seems preferable.